pkg/authorize/tollbooth/tollbooth.go

```diff
@@ -0,0 +1,124 @@
+	Reason string `json:"reason"`
+}
+
+type authorizer struct {
+	to     *url.URL
+	client *http.Client
+}
+
+func NewAuthorizer(c *http.Client, to *url.URL) *authorizer {
+	return &authorizer{
+		to:     to,
+		client: c,
+	}
+}
+
+func (a *authorizer) AuthorizeCluster(token, cluster string) (string, error) {
+	regReq := &clusterRegistration{
+		ClusterID:          cluster,
+		AuthorizationToken: token,
+	}
+
+	data, err := json.Marshal(regReq)
+	if err != nil {
+		return "", err
+	}
+
+	req, err := http.NewRequest("POST", a.to.String(), bytes.NewReader(data))
+	if err != nil {
+		return "", err
+	}
+	req.Header.Set("Content-Type", "application/json")
+	req.Header.Set("Accept", "application/json")
+	resp, err := a.client.Do(req)
+	if err != nil {
+		return "", err
+	}
+	defer func() {
+		// read the body to keep the upstream connection open
+		if resp.Body != nil {
+			io.Copy(ioutil.Discard, resp.Body)
+			resp.Body.Close()
+		}
+	}()
+	switch resp.StatusCode {
+	case http.StatusUnauthorized:
+		return "", errWithCode{error: fmt.Errorf("unauthorized"), code: http.StatusUnauthorized}
+	case http.StatusTooManyRequests:
+		return "", errWithCode{error: fmt.Errorf("rate limited, please try again later"), code: http.StatusTooManyRequests}
+	case http.StatusConflict:
+		return "", errWithCode{error: fmt.Errorf("the provided cluster identifier is already in use under a different account or is not sufficiently random"), code: http.StatusConflict}
+	case http.StatusOK, http.StatusCreated:
+		// allowed
+	default:
+		tryLogBody(resp.Body, 4*1024, fmt.Sprintf("warning: Upstream server rejected request for cluster %q with body:\n%%s", cluster))
+		return "", errWithCode{error: fmt.Errorf("upstream rejected request with code %d", resp.StatusCode), code: http.StatusInternalServerError}
+	}
+	contentType := resp.Header.Get("Content-Type")
+	mediaType, _, err := mime.ParseMediaType(contentType)
+	if err != nil || mediaType != "application/json" {
+		log.Printf("warning: Upstream server %s responded with an unknown content type %q", a.to, contentType)
+		return "", fmt.Errorf("unrecognized token response content-type %q", contentType)
+	}
+
+	regResponse, err := tryReadResponse(resp.Body, 32*1024)
+	if err != nil {
+		log.Printf("warning: Upstream server %s response could not be parsed", a.to)
+		return "", fmt.Errorf("unable to parse response body: %v", err)
+	}
+
+	if len(regResponse.AccountID) == 0 {
+		log.Printf("warning: Upstream server %s responded with an empty user string", a.to)
+		return "", fmt.Errorf("server responded with an empty user string")
+	}
+
+	return regResponse.AccountID, nil
+}
+
+func tryReadResponse(r io.Reader, limitBytes int64) (*clusterRegistration, error) {
+	body, err := ioutil.ReadAll(io.LimitReader(r, limitBytes))
+	if err != nil {
+		return nil, err
+	}
+	response := &clusterRegistration{}
+	if err := json.Unmarshal(body, response); err != nil {
+		return nil, err
+	}
+	return response, nil
+}
+
+type errWithCode struct {
```

Comment:
Nothing  ever type asserts an error to this type, we should eliminate this in a follow up

Reply:
It is used here:
https://github.com/openshift/telemeter/blob/4ca2fbb7298800d1fb37e3507928755210b84afd/pkg/authorize/jwt/handler.go#L73-L86

I am not super happy with the handling though, but we could factor this more nicely in a follow-up.